gsw_xarray/_core.py

```diff
@@ -36,55 +38,93 @@ def quantify(rv, attrs, unit_registry=None):
+    return _arg, _reg
+
+
+def pint_compat(fname, args_names, args, kwargs):
+    """
+    fname : name of the function
+    args_names : list of argument names of the function associated with *args*
+    args, kwargs : list / dict of arguments and keyword arguments given to the function
+        by the user
+    """
     if pint_xarray is None:
         return args, kwargs, None
 
-    using_pint = False
     new_args = []
     new_kwargs = {}
     registries = []
-    for arg in args:
-        if isinstance(arg, xr.DataArray):
-            if arg.pint.units is not None:
-                new_args.append(arg.pint.dequantify())
-                registries.append(arg.pint.registry)
-            else:
-                new_args.append(arg)
-        elif isinstance(arg, pint.Quantity):
-            new_args.append(arg.magnitude)
-            registries.append(arg._REGISTRY)
-        else:
-            new_args.append(arg)
+
+    for kw, arg in zip(args_names, args):
+        _arg, _reg = dequantify_reg(kw, arg)
+        new_args.append(_arg)
+        # We append registry only if kw has a unit, e.g. we skip it if kw is 'axis' or 'interp_method'
+        if input_units[kw] is not None:
+            registries.append(_reg)
 
     for kw, arg in kwargs.items():
-        if isinstance(arg, xr.DataArray):
-            if arg.pint.units is not None:
-                new_kwargs[kw] = arg.pint.dequantify()
-                registries.append(arg.pint.registry)
-            else:
-                new_kwargs[kw] = arg
-        elif isinstance(arg, pint.Quantity):
-            new_kwargs[kw] = arg.magnitude
-            registries.append(arg._REGISTRY)
-        else:
-            new_kwargs[kw] = arg
+        _arg, _reg = dequantify_reg(kw, arg)
+        new_kwargs[kw] = _arg
+        # We append registry only if kw has a unit, e.g. we skip it if kw is 'axis' or 'interp_method'
+        if input_units[kw] is not None:
+            registries.append(_reg)
 
     registries = set(registries)
+    # If there is a None in registries, but len > 1 => error
+    if (len(registries) > 1) and (None in registries):
+        raise ValueError("Mixed usage of Quantity and non Quantity is forbidden.")
+
+    try:
+        registries.remove(None)
+    except KeyError:
+        pass
+
     if len(registries) > 1:
-        raise ValueError("Quantity arguments must all belong to the same unit registry")
+        raise ValueError(
+            "Quantity arguments must all belong to the same unit registry."
+        )
     elif len(registries) == 0:
         registries = None
     else:
         (registries,) = registries
     return new_args, new_kwargs, registries
 
 
-def cf_attrs(attrs, name, check_func):
+def cf_attrs(fname, attrs, name, check_func):
     def cf_attrs_decorator(func):
         @wraps(func)
         def cf_attrs_wrapper(*args, **kwargs):
-            args, kwargs, unit_registry = pint_compat(args, kwargs)
+            args_names = get_args_names(func, args)
```

Comment:
It looks like machinery for converting everything into kwargs was being made. (the args_and_kwargs_to_args func in _function_utils.py). Should we maybe convert all inputs into a dict of kwargs and just work with that internally? I might be unaware of some edge cases, but it feels like we could get rigd of a lot of duplication in the helper functions that handle the args and kwargs seperately.

Reply:
Yes you are right, I had the same idea, and then wanted to keep the args just to stay the more general as possible. But I don't think that any of the gsw functions use args that can't be transformed into kwargs (I mean there is no gsw function defined by: `def _(arg1, arg2, *, kwargs1=None):`).

I will convert every arg to kwarg

Reply:
done